### ISSUE TITLE
Question: How to use semantic graph & add edges

### ISSUE THREAD

**Issue by dustyatx** [neuml/txtai#521]
TLDR: How do we use both similarity search and define edges with the semantic graph? 

In the example on how to build a graph it shows how to use GraphFactory to create a graph, then graph.addnode to create a node and graph.addedge to create the edges. 

However lower down in the notebook it shows how to use embeddings and emeddings.index to load data into nodes. 

Is there a way to use both of these together? I have a dataset of nodes with UID for each node and edges that link the nodes together. I'd like to use those to define the known relationships and then use similarity on the nodes to identify relationships that are inferred based on similarity. 

So for example I have a node that contains an articles of text, that node has entities which I know are mentioned in the text. They also have nodes that have categories like "Logistics". So I would want to query nodes similar to "logistics" and get "shipping", "trucking" then walk the graph to get the entities (via the linked articles)

Is this possible?

**Comment by dustyatx** [comment]
Thank you for the quick response! 

I tried that before I posted this and it wasn't working properly.. However I think I figured out what's going on.. 

When I use emeddings.index((uid, text)) it doesn't appear to use the uid as the node id in the graph it uses the index position (the order of insertion) for the node id. Since I gave each node a random UID when I generated the data, the edges aren't going to line up with the new node ids once I insert. 

when i use embeddings.graph.addnode(uid, text=content) to add a node, it does the proper UID as the node id but that  doesn't  index seem to create a vectorized version of the text so embeddings.search('my phrase') doesn't match those nodes. 

I do like the idea of being able to say some nodes need to have a 1:1 text match and others can use similarity search, that's very useful. I'm just a bit confused on how to best accomplish that. 

The other thing that is confusing.. When using the embedding approach, I have to use embeddings.index() to add at least 1 node before it will let me insert nodes using addnode(). 

I tried to initialize the graph but I'm clearly not doing that correctly since I get errors, saying "'NoneType' object has no attribute 'initialize'" 

graph = embeddings.creategraph()
graph.initialize()

**Comment by davidmezzetti** [comment]
Good point. Probably need a top level method to add graph nodes. That method would resolve the uid->internal id mapping.

I'll think about a clean implementation and follow up. I do think it's good functionality to add. 

In the meantime, if you want to get something working locally, you can look at the `embeddings.delete` method. If you use that same logic for uid lookups, you'll then be able to use internal id to add graph edges.

**Comment by dustyatx** [comment]
I looked through the embeddings.delete method but I didn't see anything obvious. I ended up adding in some logic for tracking what the uid to index position as a lookup table. Then used the lookup table to remap all the edges. Not something I'd want to do every time but good enough for a prototype. 

Appreciate the help. Hope you don't mind if I ask more questions about how to continue to iterating on my graph.

**Comment by dustyatx** [comment]
I looked through the embeddings.delete method but I didn't see anything obvious. I ended up adding in some logic for tracking what the uid to index position as a lookup table. Then used the lookup table to remap all the edges. Not something I'd want to do every time but good enough for a prototype.

Thank you for the help & this amazing project! I've been trying for over a month (failing many times) to find a database that I can build this type of graph. 

Hope you don't mind if I ask more questions about how to continue to iterating on my graph.

**Comment by davidmezzetti** [comment]
Glad you were able to get something working. 

I've added #525 to create a clean way to do this through the framework. Please take a look and see if that spec would make it easier for you to accomplish what you're trying to do.
